### PR TITLE
Add option for NPO to fit baseline first

### PR DESCRIPTION
**Pull request by ahtsan**
Refactor NPO to support either fit baseline first or fit baseline after. Also, expose functions for calculating advantages and returns for NPO variants.

### CODE REVIEW

File: src/garage/tf/algos/npo.py [hunk 1]
```diff
@@ -105,13 +107,15 @@ def __init__(self,
...
                  flatten_input=True,
+                 fit_baseline_first=False,
                  name='NPO'):
         self._name = name
         self._name_scope = tf.name_scope(self._name)
         self._use_softplus_entropy = use_softplus_entropy
         self._use_neg_logli_entropy = use_neg_logli_entropy
         self._stop_entropy_gradient = stop_entropy_gradient
         self._pg_loss = pg_loss
+        self._will_fit_baseline_first = fit_baseline_first
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: please keep member variables the same as constructor args, so they're easy to track

File: examples/tf/ppo_pendulum.py [hunk 2]
```diff
@@ -71,6 +71,7 @@ def tf_ppo_pendulum(ctxt=None, seed=1):
             entropy_method='max',
             policy_ent_coeff=0.02,
             center_adv=False,
+            fit_baseline_first=True,
```
Comment (ryanjulian) [inline examples/tf/ppo_pendulum.py]: this is hard to understand if i don't know the internal of this algorithm.

maybe you can make it a string arg named 'fit_baselin' instead, with options `'before'` and `'after'`

later we might make an in-graph baseline (that is, the baseline and policy are in the same model), in which case we could introduce the option `'joint'` or `'during'` or something.

File: src/garage/tf/algos/npo.py [hunk 3]
```diff
@@ -607,6 +639,39 @@ def _fit_baseline(self, samples_data):
         else:
             self.baseline.fit(paths)
 
+    def _get_advantages(self, baseline_var, reward_var, name):
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: what is the purpose of these two helper functions? they just call single other functions.

File: src/garage/tf/algos/npo.py [hunk 4]
```diff
@@ -562,7 +594,7 @@ def _build_entropy_term(self, i):
 
         return policy_entropy
 
-    def _fit_baseline(self, samples_data):
+    def _fit_baseline_after(self, samples_data):
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: please find a way to structure this so that the baseline fitting routine is shared between the cases.

File: src/garage/tf/algos/npo.py [hunk 5]
```diff
@@ -68,6 +68,8 @@ class NPO(BatchPolopt):
         flatten_input (bool): Whether to flatten input along the observation
             dimension. If True, for example, an observation with shape (2, 4)
             will be flattened to 8.
+        fit_baseline_first (bool): Whether to fit baseline first or after. If
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: don't forget to update the documentation of child classes

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: The "calculate returns in separate rollouts" version of NPO will need to override this method.

File: src/garage/tf/algos/npo.py (hunk 5 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: Do you mean to make this NOT optional in all the super classes?

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: Do you mean doing `fit_baseline_order=fit_baseline_order` in constructor of child classes of `NPO`, like `PPO`?

File: examples/tf/ppo_pendulum.py [hunk 6]
```diff
@@ -71,6 +71,7 @@ def tf_ppo_pendulum(ctxt=None, seed=1):
             entropy_method='max',
             policy_ent_coeff=0.02,
             center_adv=False,
+            fit_baseline_order='after',
```
Comment (ryanjulian) [inline examples/tf/ppo_pendulum.py]: please call this `fit_baseline`

File: src/garage/misc/tensor_utils.py [hunk 7]
```diff
@@ -32,11 +32,14 @@ def explained_variance_1d(ypred, y):
     Args:
         ypred (np.ndarray): Sample data from the first variable.
         y (np.ndarray): Sample data from the second variable.
+        valids (np.ndarray): Array indicating valid indices.
```
Comment (ryanjulian) [inline src/garage/misc/tensor_utils.py]: can you provide us with some information on the expected shapes and/or types of this arrays?

File: examples/tf/ppo_pendulum.py (hunk 6 above)
Comment (ahtsan) [reply examples/tf/ppo_pendulum.py]: there's already a function called `_fit_baseline`

File: src/garage/tf/algos/npo.py [hunk 8]
```diff
@@ -68,6 +67,9 @@ class NPO(BatchPolopt):
         flatten_input (bool): Whether to flatten input along the observation
             dimension. If True, for example, an observation with shape (2, 4)
             will be flattened to 8.
+        fit_baseline_order (str): Whether to fit baseline before or after,
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: this needs much better documentation. it assumes the user knows all about the internals of NPO.

Please make the point of reference the optimization step, not the prediction. In the normal case, we fit the baseline after optimizing the policy. In the strange case, we fit the baseline after optimizing the policy.

File: src/garage/tf/algos/npo.py [hunk 9]
```diff
@@ -68,6 +67,9 @@ class NPO(BatchPolopt):
...
+            either 'before' or 'after'. If 'before', baseline will be fitted
+            before making prediction. Default: 'after'.
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: Please always use complete english phrases. i.e. "Defaults to `'after'`".

File: src/garage/tf/algos/npo.py [hunk 10]
```diff
@@ -105,13 +107,19 @@ def __init__(self,
...
         self._name = name
         self._name_scope = tf.name_scope(self._name)
         self._use_softplus_entropy = use_softplus_entropy
         self._use_neg_logli_entropy = use_neg_logli_entropy
         self._stop_entropy_gradient = stop_entropy_gradient
         self._pg_loss = pg_loss
+        self._fit_baseline_order = fit_baseline_order
+
+        if fit_baseline_order not in ['before', 'after']:
+            raise ValueError('Invalid fit_baseline_order')
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: please provide more information to the user. What value did she specify, and what are the valid choices?

File: src/garage/tf/algos/npo.py [hunk 11]
```diff
@@ -192,6 +203,87 @@ def optimize_policy(self, itr, samples_data):
...
+        Args:
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        self._fit_baseline(samples_data)
+
+        # update baseline prediction for optimization
+        paths = samples_data['paths']
+        baselines = [self.baseline.predict(path) for path in paths]
```
Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: doesn't baseline have a predict_n function which does this vectorized?

Comment (ryanjulian) [inline src/garage/tf/algos/npo.py]: this is extremely strange and hard to follow. your method name says you are fitting the baseline, but then you also run the prediction step here. that's impossible to debug.

it doesn't seem like you even need different versions of `fit_baseline_before` and `fit_baseline_after` -- they both do the same thing. 

you should represent this in the logic of optimizer_policy instead. hiding it behind these methods is very confusing.

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: this version of NPO calculates returns in separate rollouts....

calculating returns across rollouts is unique to RL2 -- whatever is overriding should be in there. it makes no sense to make a special version of NPO which will only be used by RL2.

File: src/garage/tf/algos/ppo.py [hunk 12]
```diff
@@ -77,6 +81,7 @@ def __init__(self,
                  stop_entropy_gradient=False,
                  entropy_method='no_entropy',
                  flatten_input=True,
+                 fit_baseline_order='after',
```
Comment (ryanjulian) [inline src/garage/tf/algos/ppo.py]: what inheritance hierarchy do you plan for RL2/RL2PPO/RL2TRPO?

it is dangerous for this to be a constructor option in PPO when there is no reason for anyone to use it except for in RL2.

Comment (ahtsan) [reply src/garage/tf/algos/ppo.py]: I plan to have RL2PPO inherit RL2NPO, which inherits NPO by overriding `fit_baseline_before`. The conflicting part here is that if it isn't for RL2, NPO won't need the separation of `fit_baseline_before` and `fit_baseline_after`, so if you look at NPO alone separately, the refactor is weird and unnecessary.

How about we enforce PPO to have `fit_baseline_order='after'` while provides both options for RL2PPO/RL2TRPO?

File: examples/tf/ppo_pendulum.py (hunk 6 above)
Comment (ryanjulian) [reply examples/tf/ppo_pendulum.py]: then rename it. the public API is more important.

File: src/garage/tf/algos/npo.py (hunk 3 above)
Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: Do you think we should create a new class RL2NPO without inheriting NPO?

Comment (ryanjulian) [reply src/garage/tf/algos/npo.py]: if you are asking me if you should copy/paste, no.

if you can define a class RL2 which uses composition instead of inheritance for NPO, and then inherit RL2PPO and RL2TRPO from that, then yes.

take a look at Keren's MAML for what this looks like.

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: I can't use most of the code from NPO if I do composition. NPO builds the policy loss in a way that is not applicable to RL2. In order to implement "calculate returns with respect to each separate rollouts in a meta batch",  I think I can only reuse most of the code from NPO only if I inherit it and override some of the functions, especially the way it computes returns and advantages.

Comment (ahtsan) [reply src/garage/tf/algos/npo.py]: @krzentner What do you think?

Comment (krzentner) [reply src/garage/tf/algos/npo.py]: I think it's fine to have these hook point methods. Admittedly, it would be better to do something similar to MAML, but that's more difficult to do in this case while still allowing us to support variations like concatenated paths, since the related logic is spread throughout the algorithm.

File: src/garage/tf/algos/ppo.py (hunk 12 above)
Comment (ryanjulian) [reply src/garage/tf/algos/ppo.py]: see my comments above. it makes no sense to have methods `fit_baseline_before` and `fit_baseline_after` -- there is only one `fit_baseline` and what you want to control is when it's called.

I think your inheritance tree should be RL2NPO <-- NPO, RL2PPO <-- RL2NPO, RL2TRPO <-- RL2NPO. That way you can specify `fit_baseline='after'` for TRPO and PPO and allow it to be an option for RL2PPO and RL2TRPO.

**Comment by ahtsan** [comment]
I will submit another PR so it can be merged easier.
